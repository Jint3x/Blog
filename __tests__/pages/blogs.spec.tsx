/**
 * @jest-environment jsdom
 */
import ReactRenderer from "react-test-renderer"
import { screen, render } from "@testing-library/react"
import BlogPage, { FilterBlogs, DisplayBlogs, BlogPagination, getStaticProps } from "../../pages/blogs"


test("Renders the BlogPage", () => {
    let tree = ReactRenderer.create(
        <BlogPage 
         blogPosts={[]}
        />
    )
    .toJSON()

    expect(tree).toMatchSnapshot();
})


test.todo("Tests the FilterBlogs component which will be used for filtering posts by title")


test("Displays the loading state of the DisplayBlogs component", () => {
    const { container } = render(
        <DisplayBlogs 
         isLoading={true}
         blogs={[]}
        />
    )

    let isLoadingDiv = container.getElementsByClassName("animate-spinner")[0] as HTMLElement;

    expect(isLoadingDiv.nodeName).toBe("DIV");
    expect(isLoadingDiv.nextSibling || isLoadingDiv.previousSibling).toBeNull();
})


test("Renders no blogs and displays No blogs available message", () => {
    render(
        <DisplayBlogs 
         isLoading={false}
         blogs={[]}
        />
    )

    let noBlogs = screen.getByText("No blogs available");
    let parentDiv = noBlogs.parentElement as HTMLElement;

    expect(noBlogs.nodeName).toBe("H2");
    expect(parentDiv.nextSibling || parentDiv.previousSibling).toBeNull();
})


test("Renders 4 blogs", () => {
    let testBlog = {
        thumbnail: "/special/image.png",
        description: "A description",
        date: "2nd of March, 2020",
        tags: ["Experiment", "Test"],
        url: "/blogs/test"
    };

    render(
        <DisplayBlogs 
         isLoading={false}
         blogs={[{ ...testBlog, title: "A1" }, {...testBlog, title: "A2"}, {...testBlog, title: "A3"}, {...testBlog, title: "A4", description: "Custom Description"}]}
        />
    )

    let mainDescriptions = screen.getAllByText("A description");
    let customDescriptions = screen.getByText("Custom Description");

    expect(mainDescriptions.length).toBe(3);
})


test("Increases a page number", () => {
    let page = 1;
    const setPage = jest.fn((data) => data(page));

    const { rerender } = render(
        <BlogPagination 
         page={page}
         setPage={setPage}
         amountOfPosts={6}
        />
    )

    let nextPage = screen.getByText("2");
    nextPage.click();

    rerender(
        <BlogPagination 
         page={setPage?.mock?.results[0]?.value}
         setPage={setPage}
         amountOfPosts={6}
        />
    )

    let prevPage = screen.getByText("1");
    let currPage = screen.getByText("2");
    nextPage = screen.getByText("3");
    
    expect(setPage?.mock?.results[0]?.value).toBe(2);
    expect(setPage.mock.calls.length).toBe(1);
    expect(prevPage.nodeName).toBe("SPAN");
    expect(currPage.nodeName).toBe("SPAN");
    expect(nextPage.nodeName).toBe("SPAN");
})


test("Decreases a page number", () => {
    let page = 2;
    const setPage = jest.fn((data) => data(page));

    const { rerender } = render(
        <BlogPagination 
         page={page}
         setPage={setPage}
         amountOfPosts={6}
        />
    )

    let prevPage = screen.getByText("1");
    prevPage.click();

    rerender(
        <BlogPagination 
         page={page}
         setPage={setPage}
         amountOfPosts={6}
        />
    )

    screen.getByText("1");
    screen.getByText("2");

    expect(setPage?.mock?.results[0]?.value).toBe(1);
    expect(setPage.mock.calls.length).toBe(1);
})


test("Returns the correct data revalidation", async () => {
    let data = await getStaticProps();

    expect(data.revalidate).toBe(1800)
})
