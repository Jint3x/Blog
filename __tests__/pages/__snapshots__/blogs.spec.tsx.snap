// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the BlogPage 1`] = `
Array [
  <nav>
    <div
      className="
             bg-black flex pt-1 pb-2 content-center flex-col fixed w-full transition-all transform duration-1000 z-50
             sm:flex-row sm:pt-1 sm:h-10 sm:left-1/2 sm:-translate-x-1/2 sm:rounded-b-3xl sm:w-128
             
            "
    >
      <div
        className="w-full text-center sm:w-40 pl-3 pb-3"
      >
        <p
          className="text-2xl"
        >
          <span
            className="text-main-blue relative bottom-2"
          >
            Jint3x's
          </span>
          <span
            className="text-main-green relative left-1"
          >
            Blogs
          </span>
        </p>
      </div>
      <div
        className="w-full self-center sm:w-72 sm:pt-1 sm:ml-auto sm:mr-6"
      >
        <ul
          className="flex justify-around w-full"
        >
          <li
            className="border-main-green border rounded w-20 text-center text-main-blue hover:bg-main-blue-lighter transition-all cursor-pointer"
          >
            <a
              className="w-20"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            className="border-main-green border rounded w-20 text-center text-main-blue hover:bg-main-blue-lighter transition-all cursor-pointer"
          >
            <a
              className="w-20"
              href="/blogs"
            >
              Blogs
            </a>
          </li>
          <li
            className="border-main-green border rounded w-20 text-center text-main-blue hover:bg-main-blue-lighter transition-all cursor-pointer"
          >
            <a
              className="w-20"
              href="https://jint3x.dev"
            >
              About me
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>,
  <section>
    <div
      className="pt-24 pb-12 sm:pt-16"
    >
      <div>
        <h1
          className="text-center text-2xl"
        >
          Search Blogs
        </h1>
        <input
          className="w-64 sm:w-80 md:w-96 lg:w-128 h-6 sm:h-8 border pl-2 pr-2 bg-main-green-trans border-2 border-solid border-main-green rounded ml-auto mr-auto block mt-4 outline-none"
          type="text"
        />
        <p
          className="text-red-500 w-64 md:w-96 lg:w-128 ml-auto mr-auto text-center"
        >
          Search feature not available yet
        </p>
      </div>
      <section>
        <div
          className="flex sm:max-w-3xl flex-col flex-wrap justify-space-around ml-auto mr-auto xl:max-w-5xl sm:min-w-sm sm:flex sm:grid-cols-2 sm:grid-rows-0 xl:grid-cols-3 xl:grid-rows-2"
        >
          <div>
            <h2
              className="text-center mt-32 mb-32 text-xl sm:mt-64 sm:mb-64 xl:mt-80 xl:mb-80"
            >
              No blogs available
            </h2>
          </div>
        </div>
      </section>
      <div>
        <p
          className="w-12 flex justify-around ml-auto mr-auto"
        >
          <span
            className="cursor-pointer"
            onClick={[Function]}
          >
            
          </span>
          <span
            className="text-main-green"
          >
            1
          </span>
          <span
            className="cursor-pointer"
            onClick={[Function]}
          >
            
          </span>
        </p>
      </div>
    </div>
  </section>,
  <footer
    className="w-full bg-black"
  >
    <div
      className="pt-4 pb-4 flex w-56 justify-between mr-auto ml-auto sm:w-72"
    >
      <div>
        <h3
          className="text-main-green text-base"
        >
          About me
        </h3>
        <a
          className="text-main-blue text-sm no-underline pt-1 pb-1 block"
          href="https://jint3x.dev"
        >
          My portfolio
        </a>
        <a
          className="text-main-blue text-sm no-underline pt-1 pb-1 block"
          href="https://www.linkedin.com/in/jint3x/"
        >
          Linkedin
        </a>
      </div>
      <div>
        <h3
          className="text-main-green text-base"
        >
          Find me
        </h3>
        <a
          className="text-main-blue text-sm no-underline pt-1 pb-1 block"
          href="https://github.com/Jint3x"
        >
          Github
        </a>
        <a
          className="text-main-blue text-sm no-underline pt-1 pb-1 block"
          href="https://twitter.com/Jint3x"
        >
          Twitter
        </a>
      </div>
    </div>
  </footer>,
]
`;
